spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -247,36 +252,33 @@ public Table createTable(
   public CompletableFuture<WriteApiResponse> writeJsonStream(
       String tableName, InputStream jsonInputStream) {
 
-    CompletableFuture<WriteApiResponse> writeApiFutureResponse =
-        new CompletableFuture<>();
-
-    Thread asyncTask =
-        new Thread(
-            () -> {
-              try {
-                WriteApiResponse apiResponse = getWriteApiResponse(tableName, jsonInputStream);
-                writeApiFutureResponse.complete(apiResponse);
-              } catch (DescriptorValidationException | IOException e) {
-                writeApiFutureResponse.completeExceptionally(e);
-                logger.warn(String.format("Error: %s %n", e.getMessage()), e);
-              } catch (Exception e) {
-                writeApiFutureResponse.completeExceptionally(e);
-                // Restore interrupted state in case of an InterruptedException
-                Thread.currentThread().interrupt();
-              }
-            });
-    asyncTask
-        .start(); // run the thread async so that we can return the writeApiFutureResponse. This
-    // thread can be run in the ExecutorService when it has been wired-in
-
+    CompletableFuture<WriteApiResponse> writeApiFutureResponse = new CompletableFuture<>();
+    Runnable asyncTask =
+        () -> {
+          try {
+            WriteApiResponse apiResponse = getWriteApiResponse(tableName, jsonInputStream);
+            writeApiFutureResponse.complete(apiResponse);
+          } catch (Exception e) {
+            writeApiFutureResponse.completeExceptionally(e);
+            // Restore interrupted state in case of an InterruptedException
+            Thread.currentThread().interrupt();
+            logger.warn("Unable to get write API response.", e);
+          }
+        };
+    ScheduledFuture<?> asyncTaskScheduledFuture =
+        taskScheduler.schedule(asyncTask, Instant.now()); // Run this task
     // register success and failure callback
     writeApiFutureResponse.whenComplete(
         (writeApiResponse, exception) -> {
-          if (exception != null) {
-            asyncTask.interrupt(); // interrupt the thread as the developer might have cancelled the
-            // Future.
-            // This can be replaced with interrupting the ExecutorService when it has been wired-in
-            logger.info("asyncTask interrupted");
+          if (exception != null || !writeApiResponse.isSuccessful()) {
+            if (exception != null) {
+              logger.error("asyncTask interrupted", exception);
+              if (exception instanceof CancellationException) { // user have cancelled it
+                asyncTaskScheduledFuture.cancel(true);
+              }
+            } else {
+              logger.warn("Write operation failed");
+            }
             return;
           }
           logger.info("Data successfully written");
```

Comment:
Can you restructure to:
`if (exception != null) ... else if (!writeApiResponse.isSuccessful()) ... else logger.info("Data successfully written");`
?

Reply:
Done